### ISSUE TITLE
Using Custom Directive Schema in iOS client, something like @product.

### ISSUE THREAD

**Issue by javierupstack** [apollographql/apollo-ios#1664]
## Bug report

We've implemented our own Custom Directive Schema and we try use it on iOS.

## Versions

- `apollo-ios` SDK version: Up to Next Major = 0.41.0
- Xcode version: Version 12.4 (12D4e)
- Swift version: 5.0
- Package manager: SPM on  https://github.com/apollographql/apollo-ios

## Steps to reproduce

Our graphql query is working perfectly with postman:
```
query @product(key: "vpn") {
    servers {
        key name endpoint keepAlive publicKey
    }
}
```
In iOS we are using the following .graphql to solve the issue with anonymous querys.
```
query product {
    @product(key: "vpn") {
        servers {
            key name endpoint keepAlive publicKey
        }
    }
}
```
The error throw by the code gen is: `Syntax Error: Expected Name, found "@".`
We use a target to generate the code running a script:
```
"${SCRIPT_PATH}"/run-bundled-codegen.sh codegen:generate --target=swift --includes=./**/apollo_*.graphql --localSchemaFile="vpn-schema.json" VPNSchemaAPI.swift
```

## Further details
Probably we can provide more information like the schema file if you needed.
Thanks.

**Comment by designatednerd** [comment]
Right now we do not support custom directives for iOS code generation. We're hoping to be able to support it in the future, but it is not there yet. 

TBH it also looks like you're using them in a way I haven't seen before, where you're using the custom directive as the API key rather than as some sort of supplement to your code. This is a structure I'm more familiar with: 

```graphql
query product {
    product @key("vpn") {
        servers {
            key 
            name
            endpoint 
            keepAlive 
            publicKey
        }
    }
}
``` 

And I think that might be why our parser is freaking out on you.

**Comment by javiertile** [comment]
Thanks @designatednerd

**Comment by designatednerd** [comment]
Is there anything else I can help with here, or do you mind if we close this issue out at the moment since this usage is unsupported?

**Comment by crashtech** [comment]
Hey there! I'm helping with this project. How hard it would be to implement a passthrough for this directive? Its purpose is server-only. When we do `query @product(key: "vpn")` (works with mutation and subscription) it provides access to a specific subpart of the application. We thought about using a field like `product(key: "vpn") { ... }` but that would require us to have the same field implemented in the 3 operations. Do you have a recommendation for a work-around?

**Comment by designatednerd** [comment]
The biggest issue is that iOS doesn't support custom directives, so this just flat out won't work with the way we generate code.

>  We thought about using a field like product(key: "vpn") { ... } but that would require us to have the same field implemented in the 3 operations.

Can you clarify what you mean here? What three operations are you referencing?

**Comment by crashtech** [comment]
The operations are `query`, `mutation`, and `subscribe`. Since each of them has its own type, the `product` field would have to be implemented in each one of them and it would be a bit weird for mutations.

It would not be possible to make it ignore custom directives? I know that by having them implemented, Apollo can save processing by smart associating requests and performing the directives without touching the server. So, I think that a workaround would be to just throw it to the server when the request has a directive that it does not know how to handle.

**Comment by designatednerd** [comment]
It's not that simple: 

- Our parser is not recognizing the setup you're using here, it's just erroring out. 
- Even if it did parse, we generate strongly-typed code based on operations for iOS, we can't just "throw it to the server" because our codegen wouldn't understand how to generate something for it. 
- I've seriously got no idea what you're talking about with "smart associating requests" - can you explain what you mean by this? 

I think I'm missing something here when it comes to why you can't have a `product` field in a mutation, a subscription, and a query. If it's to do with the stringly-typing, why not define an `InputType`?

**Comment by designatednerd** [comment]
Ran this past some coworkers and they suggested your syntax for adding the name of the query is not correct - this should at least parse: 

```graphql
query product @product(key: "vpn") {
   servers {
       key name endpoint keepAlive publicKey
   }
}
```

**Comment by crashtech** [comment]
Yup. This is the right syntax. The name of the query is not necessary, but that is what we are trying to achieve. And that should not impact the client on anything. It is just telling the server where to get the data. Something like `@readFromDb(name: "replica")`. The client does not need to do anything about that, and the types inside are just available as described in the schema. Maybe even something like `@version(key: "123")`. This is a 100% server-only responsibility. And since the query is different, we would not have problems with caching.

**Comment by designatednerd** [comment]
The name of the query is not strictly necessary under GraphQL, but it is necessary for our code generation so we can generate a class with the proper name. 

Making that change, are you at least able to generate code?

**Comment by javierupstack** [comment]
Thanks @designatednerd the codegen script works now with:
```
query product @product(key: "vpn") {
   servers {
       key name endpoint keepAlive publicKey
   }
}
```
At least now we need to figure out another issue:
```
(lldb) po result.get()
▿ GraphQLResultError
  ▿ path : servers
    ▿ head : Optional<Node>
      ▿ some : <Node: 0x600003ba8740>
  - underlying : Apollo.JSONDecodingError.missingValue
```

**Comment by designatednerd** [comment]
Good to hear you're able to generate and use the code. I'm going to close this out since we've figured out how to get you unblocked, at least. 

The error is what you'd get back if one of the properties was not sent back from the server. I would recommend taking a look at the full `underlying` error for what is missing. If you're still having trouble, please open a new issue since that's quite a different problem than the one outlined in this issue. Thank you!
